ml_metadata/tools/mlmd_bench/thread_runner.cc

```diff
@@ -0,0 +1,120 @@
+// current workload.
+// During the execution, each operation will has a `op_stats` to record current
+// operation statistic. Each `op_stats` will be used to update the
+// `thread_stats`.
+// After the each thread has finished the execution, the workload stats will be
+// generated by merging all the thread stats inside the `thread_stats_list`. The
+// performance of the workload will be reported according to the workload stats.
+tensorflow::Status ThreadRunner::Run(Benchmark& benchmark) {
+  for (int i = 0; i < benchmark.num_workloads(); ++i) {
+    WorkloadBase* workload = benchmark.workload(i);
+    ThreadStats thread_stats_list[num_threads_];
+    std::unique_ptr<MetadataStore> set_up_store;
+    TF_RETURN_IF_ERROR(CreateMetadataStore(mlmd_config_, &set_up_store));
+    TF_RETURN_IF_ERROR(workload->SetUp(set_up_store.get()));
+    const int64 op_per_thread = workload->num_operations() / num_threads_;
+    {
+      // Create a thread pool for multi-thread execution.
+      tensorflow::thread::ThreadPool pool(tensorflow::Env::Default(),
+                                          "mlmd_bench", num_threads_);
+      // `approx_total_done` is used for reporting progress along the way.
+      int64 approx_total_done = 0;
+      for (int64 thread_index = 0; thread_index < num_threads_;
+           ++thread_index) {
+        int64 work_items_start_index = op_per_thread * thread_index;
+        ThreadStats& curr_thread_stats = thread_stats_list[thread_index];
+        pool.Schedule([this, op_per_thread, workload, work_items_start_index,
+                       &curr_thread_stats, &approx_total_done]() {
+          // Each thread uses a different MLMD client instance to talk to
+          // the same back-end.
+          std::unique_ptr<MetadataStore> store;
+          tensorflow::Status store_status =
+              CreateMetadataStore(mlmd_config_, &store);
+          // Handles abort issues for concurrent creating `store` of the db.
+          while (!store_status.ok()) {
+            store_status = CreateMetadataStore(mlmd_config_, &store);
+            // If the error is not Abort error, break the current process.
+            if (!store_status.ok() &&
+                store_status.code() != tensorflow::error::ABORTED) {
+              TF_CHECK_OK(store_status);
+            }
+          }
+          curr_thread_stats.Start();
+          // Executes the current workload by the specified
+          // `work_items_index`.
+          int64 work_items_index = work_items_start_index;
+          while (work_items_index < work_items_start_index + op_per_thread) {
+            // Each operation has a op_stats.
+            OpStats op_stats;
+            tensorflow::Status status =
+                workload->RunOp(work_items_index, store.get(), op_stats);
+            // If the error is not Abort error, break the current process.
+            if (!status.ok() && status.code() != tensorflow::error::ABORTED) {
+              TF_CHECK_OK(status);
```

Comment:
ditto

Reply:
Yes, I tried to change the TF_CHECK_OK() to TF_RETURN_IF_ERROR() inside the pool Schedule(), there will always be a Segmentation fault. I also try to print out the error message, but it seems that the program does not even come into the if statement. Wdyt? Does the error happen because we cannot do  TF_RETURN_IF_ERROR()  in a pool lambda function?